src/core/server.cc

```diff
@@ -294,150 +294,235 @@ class HTTPServiceImpl {
+
+  tensorflow::Status Stop() override;
 
  private:
-  void Handle(tfs::net_http::ServerRequestInterface* req);
-  tfs::net_http::HTTPStatusCode Health(
-      tfs::net_http::ServerRequestInterface* req,
-      const std::string& health_uri);
-  tfs::net_http::HTTPStatusCode Profile(
-      tfs::net_http::ServerRequestInterface* req,
-      const std::string& profile_uri);
-  tfs::net_http::HTTPStatusCode Infer(
-      tfs::net_http::ServerRequestInterface* req, const std::string& infer_uri);
-  tfs::net_http::HTTPStatusCode Status(
-      tfs::net_http::ServerRequestInterface* req,
-      const std::string& status_uri);
+  void Handle(evhtp_request_t* req);
+  void Health(evhtp_request_t* req, const std::string& health_uri);
+  void Profile(evhtp_request_t* req, const std::string& profile_uri);
+  void Infer(evhtp_request_t* req, const std::string& infer_uri);
+  void Status(evhtp_request_t* req, const std::string& status_uri);
+
+  static void thread_init(evhtp_t* htp, evthr_t* thread, void* arg);
+  static void thread_exit(evhtp_t* htp, evthr_t* thread, void* arg);
+
+  // Class object associated to evhtp thread, requests received are bounded
+  // with the thread that accepts it. Need to keep track of that and let the
+  // corresponding thread send back reply
+  class ThreadWiseService {
+   public:
+    class Request {
+     public:
+      Request(
+          evhtp_request_t* req, uint64_t id,
+          const std::shared_ptr<HTTPInferRequestProvider>& request_provider,
+          const std::shared_ptr<HTTPInferResponseProvider>& response_provider,
+          const std::shared_ptr<ModelInferStats>& infer_stats,
+          const std::shared_ptr<ModelInferStats::ScopedTimer>& timer)
+          : req_(req), id_(id), request_provider_(request_provider),
+            response_provider_(response_provider), infer_stats_(infer_stats),
+            timer_(timer){};
+
+     private:
+      friend class HTTPServiceImpl;
+      friend class ThreadWiseService;
+      evhtp_request_t* req_;
+      uint64_t id_;
+      RequestStatus request_status_;
+      std::shared_ptr<HTTPInferRequestProvider> request_provider_;
+      std::shared_ptr<HTTPInferResponseProvider> response_provider_;
+      std::shared_ptr<ModelInferStats> infer_stats_;
+      std::shared_ptr<ModelInferStats::ScopedTimer> timer_;
+    };
+
+    explicit ThreadWiseService(evthr_t* thread) : thread_(thread){};
+    ~ThreadWiseService() = default;
+
+    static void ReplyCallback(evthr_t* thr, void* arg, void* shared);
+
+    Request* Enqueue(
+        evhtp_request_t* req, uint64_t id,
+        const std::shared_ptr<HTTPInferRequestProvider>& request_provider,
+        const std::shared_ptr<HTTPInferResponseProvider>& response_provider,
+        const std::shared_ptr<ModelInferStats>& infer_stats,
+        const std::shared_ptr<ModelInferStats::ScopedTimer>& timer);
+
+    void FinalizeHTTPResponse(Request* req);
+
+   private:
+    std::mutex mtx_;
+    std::set<Request*> requests_;
```

Comment:
You don't need requests_ at all, do you? I don't think you need the mutex either.